src/index.ts

```diff
@@ -23,13 +39,46 @@ export {
 	Resolver,
 };
 
+export interface ResolveListeners {
+	close?: Array<() => void>;
+	data?: Array<(chunk: Buffer | string) => void>;
+	end?: Array<() => void>;
+	error?: Array<(e: Error) => void>;
+	readable?: Array<() => void>;
+	'resolved-name'?: Array<(dockerfilePath: string) => void>;
+	resolver?: Array<(resolverName: string) => void>;
+}
+
+/**
+ * Reads bundle.tarStream (typically a reference to BuildTask.buildStream) and
+ * resolves it to a Docker-compatible tar stream (i.e. selects a specified
+ * Dockerfile, translates Dockerfile.template vars and the like).
+ * @param bundle Bundle including the input tar stream
+ * @param resolvers Bundle resolvers
+ * @param resolveListeners Event listeners for tar stream resolution.
+ * Note that although you can add listeners to the returned Pack stream,
+ * the stream is piped to before it is returned and events may be missed
+ * unless the listeners are installed through the resolveListeners argument.
+ * You should always add at least an 'error' handler, or uncaught errors
+ * may crash the app.
+ * @param dockerfile User-selected Dockerfile (docker-compose `dockerfile:`)
+ */
 export function resolveInput(
 	bundle: Bundle,
 	resolvers: Resolver[],
+	resolveListeners: ResolveListeners,
 	dockerfile?: string,
-): Readable {
+): tar.Pack {
 	const extract = tar.extract();
 	const pack = tar.pack();
+	for (const event of Object.keys(resolveListeners) as Array<
```

Comment:
If I write it like that (`in` and no type cast), then on the following line I get a compiler error:
```
src/index.ts:75:21 - error TS7017: Element implicitly has an 'any' type because type 'ResolveListeners' has no index signature.
75  	const listeners = resolveListeners[event] || [];
```
Also, the `in` operator iterates over all of the object's prototype chain's properties (which in some cases results in the linter warning that the for loop needs an `if` test for "own properties"), while `Object.keys()` returns the keys for the object's own properties only.